### PR TITLE
ARROW-14150: [C++] Don't check delimiter in CSV chunker if no quoting

### PR DESCRIPTION
**Pull request by cyb70289**
Improves performance significantly for no quoting case.

### CODE REVIEW

**Comment by cyb70289** [comment]
$ archery benchmark diff --suite-filter="arrow-csv-parser-benchmark" --cc=clang-10 --cxx=clang++-10

```
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Non-regressions: (8)
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
              benchmark        baseline       contender  change %                                                                                                              counters
   ChunkCSVEscapedBlock 746.435 MiB/sec   1.325 GiB/sec    81.754       {'run_name': 'ChunkCSVEscapedBlock', 'repetitions': 0, 'repetition_index': 0, 'threads': 1, 'iterations': 3420}
 ParseCSVFlightsExample 394.605 MiB/sec 397.147 MiB/sec     0.644      {'run_name': 'ParseCSVFlightsExample', 'repetitions': 0, 'repetition_index': 0, 'threads': 1, 'iterations': 284}
    ParseCSVQuotedBlock 482.529 MiB/sec 484.527 MiB/sec     0.414        {'run_name': 'ParseCSVQuotedBlock', 'repetitions': 0, 'repetition_index': 0, 'threads': 1, 'iterations': 2088}
   ParseCSVEscapedBlock 515.472 MiB/sec 515.476 MiB/sec     0.001       {'run_name': 'ParseCSVEscapedBlock', 'repetitions': 0, 'repetition_index': 0, 'threads': 1, 'iterations': 2364}
    ChunkCSVQuotedBlock 836.720 MiB/sec 836.350 MiB/sec    -0.044        {'run_name': 'ChunkCSVQuotedBlock', 'repetitions': 0, 'repetition_index': 0, 'threads': 1, 'iterations': 3613}
ParseCSVVehiclesExample 933.527 MiB/sec 932.065 MiB/sec    -0.157      {'run_name': 'ParseCSVVehiclesExample', 'repetitions': 0, 'repetition_index': 0, 'threads': 1, 'iterations': 54}
  ParseCSVStocksExample 776.589 MiB/sec 774.898 MiB/sec    -0.218       {'run_name': 'ParseCSVStocksExample', 'repetitions': 0, 'repetition_index': 0, 'threads': 1, 'iterations': 258}
ChunkCSVNoNewlinesBlock      213.676275      211.736137    -0.908 {'run_name': 'ChunkCSVNoNewlinesBlock', 'repetitions': 0, 'repetition_index': 0, 'threads': 1, 'iterations': 3269815}
```

File: cpp/src/arrow/csv/chunker.cc [hunk 1]
```diff
@@ -116,7 +120,8 @@ class Lexer {
     if (ARROW_PREDICT_FALSE(c == '\n')) {
       goto LineEnd;
     }
-    if (ARROW_PREDICT_FALSE(c == options_.delimiter)) {
+    // treat delimiter as a normal token in quoting is disabled
```
Comment (pitrou) [inline cpp/src/arrow/csv/chunker.cc:123]: Typo: "if quoting is disabled"

**Review by pitrou — APPROVED** [review]
LGTM, thank you

File: cpp/src/arrow/csv/chunker.cc (hunk 1 above)
Comment (nealrichardson) [reply cpp/src/arrow/csv/chunker.cc:123]:
```suggestion
    // treat delimiter as a normal token if quoting is disabled
```

**Pull request closed without merging**